src/analysis/retail/druid/balance/modules/spells/RattleTheStars.tsx

```diff
@@ -0,0 +1,190 @@
+    if (event.type === EventType.FightEnd) {
+      return;
+    }
+    this.lastUpdate = event.timestamp;
+    this.lastStacks = currentStacks(event);
+  }
+
+  handleStarsurgeCast(event: CastEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.TOUCH_THE_COSMOS.id)) {
+      this.totalAPSaved +=
+        (STARSURGE_BASE_COST -
+          (this.elunedGuidanceDiscountActive ? STARSURGE_ELUNES_GUIDANCE_DISCOUNT : 0)) *
+        PERCENT_AP_SAVED_PER_STACK *
+        this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    }
+    this.buffedStarsurges[this.lastStacks] = this.buffedStarsurges[this.lastStacks] + 1;
+  }
+
+  handleStarfallCast(event: CastEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.TOUCH_THE_COSMOS.id)) {
+      this.totalAPSaved +=
+        (STARFALL_BASE_COST -
+          (this.elunedGuidanceDiscountActive ? STARFALL_ELUNES_GUIDANCE_DISCOUNT : 0)) *
+        PERCENT_AP_SAVED_PER_STACK *
+        this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    }
+    this.buffedStarfalls[this.lastStacks] = this.buffedStarfalls[this.lastStacks] + 1;
+  }
+
+  handleDamage(event: DamageEvent) {
+    const buffStacks = this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    this.totalAddedDamage += calculateEffectiveDamage(
+      event,
+      buffStacks * PERCENT_DAMAGE_INCREASE_PER_STACK,
+    );
+  }
+
+  statistic() {
+    const dpsIncrease = this.totalAddedDamage / (this.owner.fightDuration / 1000);
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.CORE(7)}
+        wide
+        size="flexible"
+        tooltip={`This approximate DPS increase of ${formatNumber(
+          dpsIncrease,
+        )} on considers the damage increase from the talent, not increased Starsurge or Starfall usages from astral power that was saved`}
+        dropdown={
+          <>
+            <table className="table table-condensed">
+              <thead>
+                <tr>
+                  <th>Stacks</th>
+                  <th>Starsurges</th>
+                  <th>Starfalls</th>
+                  <th>Uptime</th>
+                </tr>
+              </thead>
+              <tbody>
+                {this.buffStacks.map((e, i) => (
+                  <tr key={i}>
+                    <th>{i} Stacks</th>
+                    <td>{this.buffedStarsurges[i]}</td>
+                    <td>{this.buffedStarfalls[i]}</td>
+                    <td>
+                      {formatPercentage(e.reduce((a, b) => a + b, 0) / this.owner.fightDuration)}%
+                    </td>
+                  </tr>
+                ))}
+              </tbody>
+            </table>
+          </>
+        }
+      >
+        <BoringSpellValueText spellId={TALENTS_DRUID.RATTLE_THE_STARS_TALENT.id}>
+          <>
+            {formatNumber(this.totalAPSaved)} <small>Astral Power Saved</small>
+            <br />
+            {'>'}
+            {formatNumber(dpsIncrease)} <small>approx. DPS added</small>
```

Comment:
There actually is an existing module that helps format all of this and does the DPS calculation as well - you can use `<ItemPercentDamageDone greaterThan amount={this.totalAddedDamage} />`